### ISSUE TITLE
0.0.4 install failed on Ubuntu 24.04

### ISSUE THREAD

**Issue by vient** [intel/intel-lpmd#36]
`./autogen.sh prefix=/usr` reports following
```
System paths:
  prefix: /usr
  exec_prefix: NONE
  systemdunitdir: /usr/lib/systemd/system
  lpmd_binary: ${exec_prefix}/sbin/intel_lpmd
  lpmd_confdir: ${prefix}/etc/intel_lpmd
  lpmd_rundir: /intel_lpmd
```
At first I thought that `${prefix}` parts are as expected but then I got this after installation:
```
$ sudo systemctl status intel_lpmd
○ intel_lpmd.service - Intel Low Power Daemon Service
     Loaded: bad-setting (Reason: Unit intel_lpmd.service has a bad unit file setting.)
     Active: inactive (dead)

Jun 13 00:26:00 vient-laptop systemd[1]: /usr/lib/systemd/system/intel_lpmd.service:9: Neither a valid executable name nor an absolute path: ${exec_prefix}/sbin/intel_lpmd
Jun 13 00:26:00 vient-laptop systemd[1]: intel_lpmd.service: Unit configuration has fatal error, unit will not be started.
```
Service unit file indeed contain unprocessed `${...}` substitution in line `ExecStart=${exec_prefix}/sbin/intel_lpmd --systemd --dbus-enable`

Did not ever touch autotools so don't know what to do here.

**Comment by vient** [comment]
Bypassed it by manually setting some values ` ./autogen.sh prefix=/usr sbindir=/sbin sysconfdir=/etc`

**Comment by zhang-rui** [comment]
can you reproduce this after reverting commit 08b50e4c7bbe9bd50537d90ea9b621a0e973579c?

**Comment by vient** [comment]
Yes, reverting did not seem to affect those variables. After revert `./autogen.sh` still shows
```
System paths:
  prefix: NONE
  exec_prefix: NONE
  systemdunitdir: /usr/lib/systemd/system
  lpmd_binary: ${exec_prefix}/sbin/intel_lpmd
  lpmd_confdir: ${prefix}/etc/intel_lpmd
  lpmd_rundir: ${prefix}/var/run/intel_lpmd
```

**Comment by zhang-rui** [comment]
so this problem also exists in v0.0.3?

**Comment by rmeissn** [comment]
On Fedora 40, I had to use:
`./autogen.sh prefix=/usr sbindir=/usr/sbin sysconfdir=/etc rundir=/run`

**Comment by zhang-rui** [comment]
> On Fedora 40, I had to use: `./autogen.sh prefix=/usr sbindir=/usr/sbin sysconfdir=/etc rundir=/run`

does the problem also exist with tag 0.0.3 release?

**Comment by rmeissn** [comment]
Executing `./autogen.sh prefix=/` on the 0.0.3 release logs:

```autoreconf: export WARNINGS=
autoreconf: Entering directory '.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal 
autoreconf: configure.ac: tracing
autoreconf: configure.ac: not using Libtool
autoreconf: configure.ac: not using Intltool
autoreconf: running: gtkdocize --copy
autoreconf: running: /usr/bin/autoconf
autoreconf: running: /usr/bin/autoheader
autoreconf: running: automake --add-missing --copy --no-force
autoreconf: Leaving directory '.'
checking whether make supports nested variables... yes
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether to enable maintainer-specific portions of Makefiles... yes
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for gtk-doc... yes
checking for gtkdoc-check... gtkdoc-check.test
checking for gtkdoc-check... /usr/bin/gtkdoc-check
checking for gtkdoc-rebase... /usr/bin/gtkdoc-rebase
checking for gtkdoc-mkpdf... /usr/bin/gtkdoc-mkpdf
checking whether to build gtk-doc documentation... no
checking for glib-2.0 >= 2.10.0 gobject-2.0  >= 2.10.0... yes
checking for pkg-config... (cached) /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes

System paths:
  prefix: /
  exec_prefix: NONE
  systemdunitdir: /usr/lib/systemd/system
  lpmd_binary: ${exec_prefix}/sbin/intel_lpmd
  lpmd_confdir: ${prefix}/etc/intel_lpmd
  lpmd_rundir: ${prefix}/var/run/intel_lpmd

checking whether make supports the include directive... yes (GNU style)
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking for dbus_glib_global_set_disable_legacy_property_access in -ldbus-glib-1... yes
checking for dbus-1 >= 1.1 dbus-glib-1 >= 0.94... yes
checking for gio-unix-2.0 >= 2.22 gmodule-2.0... yes
checking for libxml-2.0 >= 2.4... yes
checking for libnl-3.0... yes
checking for libnl-genl-3.0... yes
checking for libsystemd... yes
checking for gdbus-codegen... /usr/bin/gdbus-codegen
checking for gcc... (cached) gcc
checking whether the compiler supports GNU C... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to enable C11 features... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking dependency style of gcc... (cached) gcc3
checking for an ANSI C-conforming const... yes
checking for inline... inline
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for size_t... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating data/Makefile
config.status: creating config.h
config.status: config.h is unchanged
config.status: executing depfiles commands
Now type 'make' to compile ..
```

**Comment by zhang-rui** [comment]
hmm, how about this? @rmeissn 
"can you reproduce this after reverting commit https://github.com/intel/intel-lpmd/commit/08b50e4c7bbe9bd50537d90ea9b621a0e973579c?"

**Comment by zhang-rui** [comment]
> On Fedora 40, I had to use: `./autogen.sh prefix=/usr sbindir=/usr/sbin sysconfdir=/etc rundir=/run`

"./autogen.sh prefix=/" does not work anymore?

**Comment by zhang-rui** [comment]
> Bypassed it by manually setting used variables `./autogen.sh prefix=/usr sbindir=/sbin sysconfdir=/etc rundir=/run`
> 
> ```
> System paths:
>   prefix: /usr
>   exec_prefix: NONE
>   systemdunitdir: /usr/lib/systemd/system
>   lpmd_binary: /sbin/intel_lpmd
>   lpmd_confdir: /etc/intel_lpmd
>   lpmd_rundir: /run/intel_lpmd
> ```

what if you use "./autogen.sh prefix=/"?

**Comment by vient** [comment]
> what if you use "./autogen.sh prefix=/"?

It was the first thing I tried, it did not work on 0.0.4.

On 0.0.5 I get this
```
System paths:
  prefix: /
  exec_prefix: NONE
  systemdunitdir: /usr/lib/systemd/system
  lpmd_binary: ${exec_prefix}/sbin/intel_lpmd
  lpmd_confdir: ${prefix}/etc/intel_lpmd
  lpmd_rundir: /intel_lpmd
```
so prefix is still not substituted in sysconfigdir used in lpmd_confdir

**Comment by zhang-rui** [comment]
I will install ubuntu 24.04 and try to reproduce this issue this week.

**Comment by zhang-rui** [comment]
@rmeissn I will also experiment with Fedora40 this week.

**Comment by zhang-rui** [comment]
Hi, all,

Here is a summary of this issue based on my research,

1. in v0.0.3, the ${exec_prefix} and ${prefix} can be seen in autogen phase it is not a problem, this is because
    a. the intel_lpmd.service is generated in "make" phase
    b. we have code to replace these strings in Makefile.am
        edit = sed \
        -e 's|@bindir[@]|$(bindir)|g' \
        -e 's|@sbindir[@]|$(sbindir)|g' \
        -e 's|@sysconfdir[@]|$(sysconfdir)|g' \
        -e 's|@localstatedir[@]|$(localstatedir)|g'
    c. the final intel_lpmd.service file is generated correctly

2. in v0.0.4/v0.0.5, the generation of intel_lpmd.service is moved to autogen phase, as you can see in
              commit d66428a04777fcf0615ec14e0d3c347106e3a6ea
              Author:     Tomasz Kłoczko <kloczek@github.com>
              AuthorDate: Sat Feb 10 13:46:53 2024 +0000
              Commit:     Tomasz Kłoczko <kloczek@github.com>
              CommitDate: Sat Feb 10 13:46:53 2024 +0000
              
                  automake and auto conf improvements
                  
                  - remove not used `include $(GLIB_MAKEFILE)` lines
                  - use automake to build intel_lpmd_control
                  - **use autoconf AC_CONFIG_FILES() to generate data/intel_lpmd.service and
                    data/org.freedesktop.intel_lpmd.service instead custom sed rules**
                  - more [] quotes in configure ac
                  - use $rundir and bumped minimum autoconf version required to 2.70
                  - remove use AM_CFLAGS in main Makefile.am
                  - remove . from SUBDIRS in main Makefile.am
                  - use PKG_CHECK_MODULES() to detect dbus-glib-1
                  - remove redundant AC_SUBST(FOO_CFLAGS) and AC_SUBST(FOO_LIBS)
                    (PKG_CHECK_MODULES() adds that automatically)
                  
                  Signed-off-by: Tomasz Kłoczko <kloczek@github.com>
However, I think there is a bug in this code, causing that the ${exec_prefix} and ${prefix} are not replaced correctly. and we see broken intel_lpmd.service finally.
                  [Unit]
                  Description= Intel Low Power Daemon Service
                  ConditionVirtualization=no
                  
                  [Service]
                  Type=dbus
                  SuccessExitStatus=2
                  BusName=org.freedesktop.intel_lpmd
                  **ExecStart=${exec_prefix}/sbin/intel_lpmd --systemd --dbus-enable**
                  Restart=on-failure
                  
                  [Install]
                  WantedBy=multi-user.target
                  Alias=org.freedesktop.intel_lpmd.service


So, in short, this is a v0.0.4 regression. You can apply below process with both v0.0.3 and v0.0.4/v0.0.5 to confirm this.
1. run autogen.sh and ignore its output
2. run "make" and "sudo make install"
3. run "sudo systemctl status intel_lpmd" and check the intel_lpmd service status

**Comment by zhang-rui** [comment]
@kloczek, please kindly take a check of this issue.
IMO, it is a regression introduced by commit https://github.com/intel/intel-lpmd/commit/d66428a04777fcf0615ec14e0d3c347106e3a6ea

**Comment by kloczek** [comment]
Just retested 0.0.5 and all look OK now.
Please close this ticket (I cannot do that).
Thank you.

**Comment by zhang-rui** [comment]
@kloczek,
can you please attach
1. the out of "sudo systemctl status intel_lpmd"
2. and the content of intel_lpmd.service

and what distro and what autogen.sh command do you use?

**Comment by zhang-rui** [comment]
I can't close this right now because I can reproduce the issue as I described in ubuntu 22.04.
This is the intel_lpmd.service generated by v0.0.5
      [Unit]
      Description= Intel Low Power Daemon Service
      ConditionVirtualization=no
      
      [Service]
      Type=dbus
      SuccessExitStatus=2
      BusName=org.freedesktop.intel_lpmd
      ExecStart=${exec_prefix}/sbin/intel_lpmd --systemd --dbus-enable
      Restart=on-failure
      
      [Install]
      WantedBy=multi-user.target
      Alias=org.freedesktop.intel_lpmd.service

And this is the intel_lpmd.service file generated by v0.0.3

      [Unit]
      Description= Intel Low Power Daemon Service
      ConditionVirtualization=no
      
      [Service]
      Type=dbus
      SuccessExitStatus=2
      BusName=org.freedesktop.intel_lpmd
      ExecStart=/usr/local/sbin/intel_lpmd --systemd --dbus-enable
      Restart=on-failure
      
      [Install]
      WantedBy=multi-user.target
      Alias=org.freedesktop.intel_lpmd.service

**Comment by kloczek** [comment]
> @kloczek, can you please attach
> 
> 1. the out of "sudo systemctl status intel_lpmd"
> 2. and the content of intel_lpmd.service

Reported issue was about testing after build.

> and what distro and what autogen.sh command do you use?

I'm using my own distro.

**Comment by rmeissn** [comment]
This issue is fixed for me with v0.0.6 on Fedora 40.

**Comment by zhang-rui** [comment]
Problem fixed in latest release.

**Comment by Flashwalker** [comment]
configure as `./autogen.sh prefix=/usr --localstatedir=/var --sysconfdir=/etc` works for me on Ubuntu 24.04

**Comment by Flashwalker** [comment]
> can you please attach
> 
> 1. the out of "sudo systemctl status intel_lpmd"
> 2. and the content of intel_lpmd.service

I;m just installed from git, enabled and rebooted and the service is always dead:

<details><summary>sudo systemctl status intel_lpmd</summary>

```
○ intel_lpmd.service - Intel Linux Energy Optimizer (lpmd) Service
     Loaded: loaded (/usr/lib/systemd/system/intel_lpmd.service; enabled; preset: enabled)
     Active: inactive (dead) since Wed 2025-05-28 22:19:12 +07; 2min 27s ago
       Docs: man:intel_lpmd(8)
    Process: 12006 ExecStart=/usr/sbin/intel_lpmd --systemd --dbus-enable (code=exited, status=2)
   Main PID: 12006 (code=exited, status=2)
        CPU: 13ms

мая 28 22:19:12 redtop systemd[1]: Starting intel_lpmd.service - Intel Linux Energy Optimizer (lpmd) Service...
мая 28 22:19:12 redtop systemd[1]: intel_lpmd.service: Deactivated successfully.
мая 28 22:19:12 redtop systemd[1]: Started intel_lpmd.service - Intel Linux Energy Optimizer (lpmd) Service.

```

</details> 

<details><summary>files installed</summary>

/etc/dbus-1/system.d/org.freedesktop.intel_lpmd.conf

/etc/intel_lpmd/intel_lpmd_config.xml
/etc/intel_lpmd/intel_lpmd_config_F6_M170.xml
/etc/intel_lpmd/intel_lpmd_config_F6_M189.xml

/usr/bin/intel_lpmd_control

/usr/lib/systemd/system/intel_lpmd.service

/usr/sbin/intel_lpmd

/usr/share/dbus-1/system-services
/usr/share/dbus-1/system-services/org.freedesktop.intel_lpmd.service

/usr/share/doc/intel-lpmd/AUTHORS
/usr/share/doc/intel-lpmd/COPYING
/usr/share/doc/intel-lpmd/ChangeLog
/usr/share/doc/intel-lpmd/NEWS
/usr/share/doc/intel-lpmd/README.md
/usr/share/doc/intel-lpmd/doc/WLT_proxy.md

/usr/share/man/man5/intel_lpmd_config.xml.5.gz
/usr/share/man/man8/intel_lpmd.8.gz
/usr/share/man/man8/intel_lpmd_control.8.gz

/var/cache/man/CACHEDIR.TAG
/var/cache/man/ca/index.db
/var/cache/man/ca@valencia/index.db
/var/cache/man/cs/index.db
/var/cache/man/da/index.db
/var/cache/man/de/index.db
/var/cache/man/el/index.db
/var/cache/man/es/index.db
/var/cache/man/et/index.db
/var/cache/man/fi/index.db
/var/cache/man/fr/index.db
/var/cache/man/fr.ISO8859-1/index.db
/var/cache/man/fr.UTF-8/index.db
/var/cache/man/gl/index.db
/var/cache/man/hr/index.db
/var/cache/man/hu/index.db
/var/cache/man/id/index.db
/var/cache/man/index.db
/var/cache/man/it/index.db
/var/cache/man/ja/index.db
/var/cache/man/ko/index.db
/var/cache/man/nb/index.db
/var/cache/man/nl/index.db
/var/cache/man/pl/index.db
/var/cache/man/pt/index.db
/var/cache/man/pt_BR/index.db
/var/cache/man/ro/index.db
/var/cache/man/ru/index.db
/var/cache/man/sk/index.db
/var/cache/man/sl/index.db
/var/cache/man/sr/index.db
/var/cache/man/sr@latin/index.db
/var/cache/man/sv/index.db
/var/cache/man/tr/index.db
/var/cache/man/uk/index.db
/var/cache/man/vi/index.db
/var/cache/man/zh/index.db
/var/cache/man/zh_CN/index.db
/var/cache/man/zh_TW/index.db

</details>
